semantic-core/src/Data/Core/Parser.hs

```diff
@@ -48,65 +50,79 @@ core :: (TokenParsing m, Monad m) => m (Term Core User)
+application = projection `chainl1` (pure (Core.$$))
+
+projection :: (TokenParsing m, Monad m) => m (Term Core User)
+projection = foldl' (Core....) <$> atom <*> many (namedValue <$ dot <*> name)
 
 atom :: (TokenParsing m, Monad m) => m (Term Core User)
 atom = choice
   [ comp
-  , ifthenelse
-  , edge
   , lit
   , ident
-  , assign
   , parens expr
   ]
 
 comp :: (TokenParsing m, Monad m) => m (Term Core User)
-comp = braces (Core.block <$> sepEndByNonEmpty expr semi) <?> "compound statement"
+comp = braces (Core.do' <$> sepEndByNonEmpty statement semi) <?> "compound statement"
+
+statement :: (TokenParsing m, Monad m) => m (Maybe (Named User) Core.:<- Term Core User)
+statement
+  =   try ((Core.:<-) . Just <$> name <* symbol "<-" <*> expr)
+  <|> (Nothing Core.:<-) <$> expr
+  <?> "statement"
 
 ifthenelse :: (TokenParsing m, Monad m) => m (Term Core User)
 ifthenelse = Core.if'
-  <$ reserved "if"   <*> core
-  <* reserved "then" <*> core
-  <* reserved "else" <*> core
+  <$ reserved "if"   <*> expr
+  <* reserved "then" <*> expr
+  <* reserved "else" <*> expr
   <?> "if-then-else statement"
 
-assign :: (TokenParsing m, Monad m) => m (Term Core User)
-assign = (Core..=) <$> try (lvalue <* symbolic '=') <*> core <?> "assignment"
+rec :: (TokenParsing m, Monad m) => m (Term Core User)
+rec = Core.rec <$ reserved "rec" <*> name <* symbolic '=' <*> expr <?> "recursive binding"
 
-edge :: (TokenParsing m, Monad m) => m (Term Core User)
-edge = kw <*> expr where kw = choice [ Core.edge Lexical <$ reserved "lexical"
-                                     , Core.edge Import  <$ reserved "import"
-                                     , Core.load         <$ reserved "load"
-                                     ]
+load :: (TokenParsing m, Monad m) => m (Term Core User)
+load = Core.load <$ reserved "load" <*> expr
 
 lvalue :: (TokenParsing m, Monad m) => m (Term Core User)
 lvalue = choice
-  [ Core.let' . namedValue <$ reserved "let" <*> name
+  [ projection
   , ident
   , parens expr
   ]
 
 -- * Literals
 
 name :: (TokenParsing m, Monad m) => m (Named User)
-name = named' <$> identifier <?> "name" where
+name = named' <$> identifier <?> "name"
 
 lit :: (TokenParsing m, Monad m) => m (Term Core User)
 lit = let x `given` n = x <$ reserved n in choice
   [ Core.bool True  `given` "#true"
   , Core.bool False `given` "#false"
   , Core.unit       `given` "#unit"
-  , Core.frame      `given` "#frame"
-  , between (string "\"") (string "\"") (Core.string . fromString <$> many ('"' <$ string "\\\"" <|> noneOf "\""))
-  , lambda
+  , record
+  , token (between (string "\"") (string "\"") (Core.string . fromString <$> many (escape <|> (noneOf "\"" <?> "non-escaped character"))))
   ] <?> "literal"
+  where escape = char '\\' *> choice
+          [ '"'  <$ string "\""
+          , '\n' <$ string "n"
+          , '\r' <$ string "r"
+          , '\t' <$ string "t"
+          ] <?> "escape sequence"
```

Comment:
We’re probably going to want to handle more escape sequences in future, but this is a good start for now.